File: Platforms/Common/Reign_Video_D3D11_Component/Shaders/PixelShader.cpp
#include "pch.h"
#include "PixelShader.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	PixelShaderErrors PixelShaderCom::Init(VideoCom^ video, ShaderModelCom^ shaderModel)
	{
		null();
		this->video = video;
		resources = shaderModel->resources;
		variableBuffer = shaderModel->variableBuffer;
		resourceKnownCount = shaderModel->resourceKnownCount;

		ID3D11PixelShader* shaderTEMP;
		#if WINDOWS
		ID3DBlob* codeBlob = shaderModel->code;
		if (FAILED(video->device->CreatePixelShader(codeBlob->GetBufferPointer(), codeBlob->GetBufferSize(), NULL, &shaderTEMP)))
		#else
		if (FAILED(video->device->CreatePixelShader(shaderModel->code, shaderModel->codeSize, NULL, &shaderTEMP)))
		#endif
		{
			return PixelShaderErrors::PixelShader;
		}
		shader = shaderTEMP;

		return PixelShaderErrors::None;
	}

	PixelShaderCom::~PixelShaderCom()
	{
		if (shader) shader->Release();
		null();
	}

	void PixelShaderCom::null()
	{
		shader = 0;
	}
	#pragma endregion

	#pragma region Methods
	void PixelShaderCom::Apply()
	{
		if (variableBuffer)
		{
			ID3D11Buffer* buffer = variableBuffer;
			video->deviceContext->PSSetConstantBuffers(0, 1, &buffer);
		}

		if (resourceKnownCount != 0 && resources)
		{
			video->deviceContext->PSSetShaderResources(0, resourceKnownCount, resources);
		}

		video->deviceContext->PSSetShader(shader, NULL, 0);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/Shaders/PixelShader.cs
﻿using System;
using Reign.Core;
using Reign.Video;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class PixelShader : ShaderModel
	{
		#region Properties
		internal PixelShaderCom pixelShaderCom;
		#endregion

		#region Constructors
		#if WIN32
		public PixelShader(Shader shader, string code, ShaderVersions shaderVersion)
		: base(shader, code, ShaderTypes.PS, shaderVersion)
		#else
		public PixelShader(Shader shader, byte[] code)
		: base(shader, code, ShaderTypes.PS)
		#endif
		{
			try
			{
				var video = shader.FindParentOrSelfWithException<Video>();
				pixelShaderCom = new PixelShaderCom();
				var error = pixelShaderCom.Init(video.com, com);

				if (error == PixelShaderErrors.PixelShader) Debug.ThrowError("PixelShader", "Failed to create PixelShader");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (pixelShaderCom != null)
			{
				pixelShaderCom.Dispose();
				pixelShaderCom = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public override void Apply()
		{
			base.Apply();
			pixelShaderCom.Apply();
		}

		#if WINRT || WP8
		public override int Variable(string name)
		{
			for (int i = 0; i != shader.psVariableNames.Count; ++i)
			{
				if (shader.psVariableNames[i] == name) return shader.psVariableByteOffsets[i];
			}

			return -1;
		}

		public override int Resource(string name)
		{
			for (int i = 0; i != shader.psResourceCount; ++i)
			{
				if (shader.psResourceNames[i] == name) return shader.psResourceIndices[i];
			}

			return -1;
		}
		#endif
		#endregion
	}
}
